src/strongbox/core.clj

```diff
@@ -767,196 +658,114 @@
-          catalogue (query-db :addon-summary-list)
-
-          match-results (-db-match-installed-addons-with-catalogue inst-addons)
-          [matched unmatched] (utils/split-filter #(contains? % :final) match-results)
-
-          matched (mapv :final matched)
-
-          unmatched-names (set (map :name unmatched))
-
-          expanded-installed-addon-list (into matched unmatched)
-
-          [num-installed num-matched] [(count inst-addons) (count matched)]]
-
-      (when-not (= num-installed num-matched)
-        (info "num installed" num-installed ", num matched" num-matched))
-
-      (when-not (empty? unmatched)
-        (warn "you need to manually search for them and then re-install them")
-        (warn (format "failed to find %s installed addons in the '%s' catalogue: %s"
-                      (count unmatched)
-                      (name (get-state :cfg :selected-catalogue))
-                      (clojure.string/join ", " unmatched-names))))
-
-      (update-installed-addon-list! expanded-installed-addon-list))))
+  ([]
+   (when (selected-addon-dir) ;; don't even bother if we have nothing to match it to
+     (match-installed-addons-with-catalogue (get-state :db) (get-state :installed-addon-list))))
+  ([database installed-addon-list]
+   (info "matching installed addons to catalogue")
+   (let [match-results (db/-db-match-installed-addons-with-catalogue (get-state :db) installed-addon-list)
+         [matched unmatched] (utils/split-filter :matched? match-results)
+
+         ;; for those that *did* match, merge the installed addon data together with the catalogue data
+         matched (mapv #(moosh-addons (:installed-addon %) (:catalogue-match %)) matched)
+         ;; and then make them a single list of addons again
+         expanded-installed-addon-list (into matched unmatched)
+
+         ;; todo: metrics gathering is good, but this is a little adhoc.
+         ;; some metrics we'll emit for the user
+         [num-installed num-matched] [(count installed-addon-list) (count matched)]
+         unmatched-names (set (map :name unmatched))]
+
+     (when-not (= num-installed num-matched)
+       (info "num installed" num-installed ", num matched" num-matched))
+
+     (when-not (empty? unmatched)
+       (warn "you need to manually search for them and then re-install them")
+       (warn (format "failed to find %s installed addons in the '%s' catalogue: %s"
+                     (count unmatched)
+                     (name (get-state :cfg :selected-catalogue))
+                     (clojure.string/join ", " unmatched-names))))
+
+     (update-installed-addon-list! expanded-installed-addon-list))))
 
 ;; catalogue db handling
 
-(defn db-shutdown
+(defn query-db
+  [query-kw & [arg-list]]
+  (db/stored-query (get-state :db) query-kw arg-list))
+
+;; todo: should I distinguish between not-set (nil) and empty ([])?
+;; it is possible to have an empty catalogue or no catalogue ...
```

Comment:
do!! it may be faster now, but we're still re-reading the catalogue and re-loading the database on each refresh.